File: examples/bun-app/src/handlers/microservices.ts
/**
 * Microservices user registration step handlers.
 *
 * 5 steps forming a diamond dependency pattern:
 *   CreateUserAccount
 *        |-- SetupBillingProfile  --|
 *        |-- InitializePreferences -|
 *                                    |-- SendWelcomeSequence -> UpdateUserStatus
 *
 * Thin DSL wrappers that delegate to ../services/microservices for business logic.
 */

import { defineHandler, PermanentError } from '@tasker-systems/tasker';
import { CreateUserAccountInputSchema } from '../services/schemas';
import * as svc from '../services/microservices';

export const CreateUserHandler = defineHandler(
  'Microservices.StepHandlers.CreateUserHandler',
  { inputs: { email: 'email', username: 'username', plan: 'plan', metadata: 'metadata' } },
  async ({ email, username, plan, metadata }) => {
    const parsed = CreateUserAccountInputSchema.safeParse({
      email,
      username,
      plan: plan || undefined,
      metadata: metadata || undefined,
    });

    if (!parsed.success) {
      const fields = parsed.error.issues.map((i) => i.path.join('.')).join(', ');
      throw new PermanentError(
        `Input validation failed: ${fields} — ${parsed.error.issues.map((i) => i.message).join('; ')}`,
      );
    }

    return svc.createUserAccount(parsed.data);
  },
);

export const SetupBillingHandler = defineHandler(
  'Microservices.StepHandlers.SetupBillingHandler',
  { depends: { userData: 'create_user_account' } },
  async ({ userData }) => svc.setupBilling(userData as Record<string, unknown>),
);

export const InitPreferencesHandler = defineHandler(
  'Microservices.StepHandlers.InitPreferencesHandler',
  {
    depends: { userData: 'create_user_account' },
    inputs: { metadata: 'metadata' },
  },
  async ({ userData, metadata }) =>
    svc.initPreferences(
      userData as Record<string, unknown>,
      metadata as Record<string, unknown> | undefined,
    ),
);

export const SendWelcomeHandler = defineHandler(
  'Microservices.StepHandlers.SendWelcomeHandler',
  {
    depends: {
      userData: 'create_user_account',
      billingData: 'setup_billing_profile',
      preferencesData: 'initialize_preferences',
    },
  },
  async ({ userData, billingData, preferencesData }) =>
    svc.sendWelcome(
      userData as Record<string, unknown>,
      billingData as Record<string, unknown>,
      preferencesData as Record<string, unknown>,
    ),
);

export const UpdateStatusHandler = defineHandler(
  'Microservices.StepHandlers.UpdateStatusHandler',
  {
    depends: {
      userData: 'create_user_account',
      billingData: 'setup_billing_profile',
      preferencesData: 'initialize_preferences',
      welcomeData: 'send_welcome_sequence',
    },
  },
  async ({ userData, billingData, preferencesData, welcomeData }) =>
    svc.updateStatus(
      userData as Record<string, unknown>,
      billingData as Record<string, unknown>,
      preferencesData as Record<string, unknown>,
      welcomeData as Record<string, unknown>,
    ),
);

File: examples/fastapi-app/app/handlers/microservices.py
"""Microservices user registration step handlers.

5 steps forming a diamond dependency pattern:
  CreateUserAccount
       |-- SetupBillingProfile  --|
       |-- InitializePreferences -|
                                   |-- SendWelcomeSequence -> UpdateUserStatus

Thin DSL wrappers that delegate to app.services.microservices for business logic.
"""

from __future__ import annotations

from tasker_core.step_handler.functional import depends_on, inputs, step_handler
from tasker_core.types import StepContext

from app.services import microservices as svc
from app.services.types import (
    CreateUserAccountInput,
    MicroservicesCreateUserResult,
    MicroservicesInitPreferencesResult,
    MicroservicesSendWelcomeResult,
    MicroservicesSetupBillingResult,
)


@step_handler("create_user_account")
@inputs(CreateUserAccountInput)
def create_user_account(inputs: CreateUserAccountInput, context: StepContext):
    # Input validation (required fields) is handled by the model's
    # @model_validator — see CreateUserAccountInput in app/services/types.py.
    return svc.create_user_account(inputs)


@step_handler("setup_billing_profile")
@depends_on(user_data=("create_user_account", MicroservicesCreateUserResult))
def setup_billing_profile(user_data: MicroservicesCreateUserResult, context: StepContext):
    return svc.setup_billing_profile(user_data=user_data)


@step_handler("initialize_preferences")
@depends_on(user_data=("create_user_account", MicroservicesCreateUserResult))
@inputs(CreateUserAccountInput)
def initialize_preferences(
    user_data: MicroservicesCreateUserResult,
    inputs: CreateUserAccountInput,
    context: StepContext,
):
    return svc.initialize_preferences(user_data=user_data, custom_prefs=inputs.preferences)


@step_handler("send_welcome_sequence")
@depends_on(
    user_data=("create_user_account", MicroservicesCreateUserResult),
    billing_data=("setup_billing_profile", MicroservicesSetupBillingResult),
    prefs_data=("initialize_preferences", MicroservicesInitPreferencesResult),
)
def send_welcome_sequence(
    user_data: MicroservicesCreateUserResult,
    billing_data: MicroservicesSetupBillingResult,
    prefs_data: MicroservicesInitPreferencesResult,
    context: StepContext,
):
    return svc.send_welcome_sequence(
        user_data=user_data, billing_data=billing_data, prefs_data=prefs_data
    )


@step_handler("update_user_status")
@depends_on(
    user_data=("create_user_account", MicroservicesCreateUserResult),
    billing_data=("setup_billing_profile", MicroservicesSetupBillingResult),
    preferences_data=("initialize_preferences", MicroservicesInitPreferencesResult),
    welcome_data=("send_welcome_sequence", MicroservicesSendWelcomeResult),
)
def update_user_status(
    user_data: MicroservicesCreateUserResult,
    billing_data: MicroservicesSetupBillingResult,
    preferences_data: MicroservicesInitPreferencesResult,
    welcome_data: MicroservicesSendWelcomeResult,
    context: StepContext,
):
    return svc.update_user_status(
        user_data=user_data,
        billing_data=billing_data,
        preferences_data=preferences_data,
        welcome_data=welcome_data,
    )
